metricflow-semantics/metricflow_semantics/model/semantics/linkable_element_set_base.py

```diff
@@ -0,0 +1,257 @@
+
+    This class will be updated / renamed.
+    """
+
+    @abstractmethod
+    def filter(
+        self,
+        element_filter: LinkableElementFilter,
+    ) -> Self:
+        """Filter elements in the set.
+
+        First, only elements with at least one property in the "with_any_of" set are retained. Then, any elements with
+        a property in "without_any_of" set are removed. Lastly, any elements with all properties in without_all_of
+        are removed.
+        """
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def is_empty(self) -> bool:
+        """Whether this maps to an empty set of specs.
+
+        If `.specs` is created lazily, there are cases where the caller only needs to know that there are matching
+        specs.
+        """
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def specs(self) -> Sequence[LinkableInstanceSpec]:
+        """Converts the items represented by this to the corresponding spec objects."""
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def annotated_specs(self) -> Sequence[AnnotatedSpec]:
+        """Return the annotated specs represented by this."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def filter_by_spec_patterns(self, spec_patterns: Sequence[SpecPattern]) -> Self:
+        """Filter the elements in the set by the given spec patters.
+
+        Returns a new set consisting of the elements in the `LinkableElementSet` that have a corresponding spec that
+        match all the given spec patterns.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def intersection(self, *others: Self) -> Self:
+        """Intersect this set with the other set to determine possible group-by items."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def union(self, *others: Self) -> Self:
+        """Intersect this set with the other set to determine possible group-by items."""
+        raise NotImplementedError
+
+
+@fast_frozen_dataclass()
+class AnnotatedSpec(SerializableDataclass):
+    """Groups a spec with more context about the spec.
+
+    * This is intended to consolidate the `LinkableElement` classes during migration.
+    * This class will be updated / renamed.
+    """
+
+    element_type: LinkableElementType
+    element_name: str
+    entity_link_names: Tuple[str, ...]
+    time_grain: Optional[ExpandedTimeGranularity]
+    date_part: Optional[DatePart]
+    metric_subquery_entity_link_names: Tuple[str, ...]
+
+    element_properties: Tuple[LinkableElementProperty, ...]
+    # The semantic model(s) where the element (e.g. the categorical dimension) was defined.
+    # There can be multiple models if it's a metric / derived metric that references multiple measures, and the join
+    # path from the measure to the dimension is different.
+    origin_semantic_model_names: Tuple[str, ...]
+    source_semantic_model_names: Tuple[str, ...]
```

Comment:
Can we better distinguish these two properties? From the names, they sound like the same thing.

If these are going away shortly anyway, though, may not be worth the effort.

Reply:
Any ideas? This was the best I got so far.

Reply:
Could keep the names that populated them! So then one of them would be `derived_from_semantic_models`, at least for consistency.
For something more clear but verbose, could be `origin_and_joined_semantic_models`.